### PR TITLE
add AWS VPC and EKS reference resources for Karpenter

### PR DESCRIPTION
**Pull request by lpm0073**
For [issue #7](https://github.com/openedx/openedx-k8s-harmony/issues/7): refactor the new infra-example folder into infra-examples, and add Terraform modules for AWS reference resources.

Adds generic AWS Virtual Private Cloud and Elastic Kubernetes Service modules, both of which are preconfigured as necessary to support use of [Karpenter](https://karpenter.sh/) for node auto-scaling and pod bin packing.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @lpm0073!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by e0d** [comment]
@lpm0073 can you amend the commit messages to follow our conventional commit standard? https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html#type

**Comment by itsjeyd** [comment]
@lpm0073 Thank you for your contribution!

In addition to changing the commit messages as @e0d mentioned above, it would be great if you could bring your branch up-to-date with the `main` branch.

Once that's done this PR will be ready for engineering review.

**Comment by itsjeyd** [comment]
Hey @lpm0073, a friendly reminder to follow up on the comments above.

**Comment by gabor-boros** [comment]
@bradenmacdonald If I understand correctly, I would need to be a core contributor to review this, which would require a 20h commitment , that I have concerns with from Serenity point of view at the moment. Ref: https://github.com/openedx/openedx-k8s-harmony/issues/25#issuecomment-1477716458

**Comment by bradenmacdonald** [comment]
@lpm0073 said on today's call that he is still finding a way to refactor this, to improve the design/code.

**Comment by antoviaque** [comment]
@gabor-boros +1 to what @bradenmacdonald said - you don't need to be a core contributor to do the review. Could you schedule the review when it's ready?

**Comment by lpm0073** [comment]
The code itself is final. I was unsatisfied with alternative approaches for various reasons. @e0d asked me to modify the commit comments so that these conform to a tCRIL standard. This is still pending. I'll take care of this later today.

**Comment by itsjeyd** [comment]
Hey @lpm0073, it looks like this PR will need a rebase. Once that's done and the commit linter check passes we'll line it up for engineering review.

File: .gitignore
Comment (gabor-boros) [inline .gitignore:13]: Why do we need these notes file in the first place?

File: README.md
Comment (gabor-boros) [inline README.md:158]: nit: Don't you think it would be more straightforward to call it `infra-examples/digitalocean`?

File: README.md
Comment (gabor-boros) [inline README.md:162]: Based on the above change, shouldn't this be `export KUBECONFIG=`pwd`/infra-examples/digitalocean/kubeconfig`? Or would we keep it as is?

File: infra-examples/aws/k8s-cluster/README.rst
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/README.rst:1]: I'm not agains reStructuredText, but shouldn't we uniformly use markdown for documentation?

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:21]:
```suggestion
# Create the IAM role.
```

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:45]:
```suggestion
# Attach the required AWS managed policy to the role
```

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:52]:
```suggestion
# Restart the ebs-csi-controller deployment for the annotation to take effect
```

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:54]: It is absolutely just about the blank line.
```suggestion
```

File: infra-examples/aws/k8s-cluster/main.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/main.tf:53]: Couldn't shouldn't we use the [Session Manager](https://docs.aws.amazon.com/prescriptive-guidance/latest/patterns/install-ssm-agent-on-amazon-eks-worker-nodes-by-using-kubernetes-daemonset.html) for this? Or did I misunderstood something here?

File: infra-examples/aws/k8s-cluster/main.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/main.tf:84]: I wasn't able to find the latest versions, but are these the latest versions? From the [docs](https://docs.aws.amazon.com/eks/latest/userguide/eks-add-ons.html) I would say not, but again, I didn't find the latest versions.

File: infra-examples/aws/k8s-cluster/main.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/main.tf:197]:
```suggestion
```

File: infra-examples/aws/k8s-cluster/main.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/main.tf:217]:
```suggestion
```

File: infra-examples/aws/k8s-cluster/variables.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/variables.tf:42]:
```suggestion
  default     = "1.26"
```

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/versions.tf:16]:
```suggestion
      version = "~> 3.5"
```

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/versions.tf:21]:
```suggestion
      version = "~> 4.65"
```

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/versions.tf:29]:
```suggestion
      version = "~> 2.9"
```

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/versions.tf:33]:
```suggestion
      version = "~> 2.20"
```

File: infra-examples/aws/main.tf
Comment (gabor-boros) [inline infra-examples/aws/main.tf:1]: This should be removed or fixed.

File: infra-examples/aws/vpc/main.tf
Comment (gabor-boros) [inline infra-examples/aws/vpc/main.tf:50]:
```suggestion
  version                = "~> 4.0"
```

File: infra-examples/aws/vpc/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/vpc/versions.tf:15]:
```suggestion
      version = "~> 4.65"
```

File: infra-examples/aws/vpc/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/vpc/versions.tf:19]:
```suggestion
      version = "~> 2.4"
```

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (gabor-boros) [inline infra-examples/aws/k8s-cluster/versions.tf:13]:
```suggestion
    local = {
      source  = "hashicorp/local"
      version = "~> 2.4"
     }
```

**Review by gabor-boros — CHANGES_REQUESTED** [review]
@lpm0073 I just checked the PR. Very nice work on this!

I mostly have suggestions to bump version numbers, questions about using SSM and some smaller things.

**Comment by itsjeyd** [comment]
@lpm0073 It looks like there's three things to follow up on:

* Comments from @gabor-boros's [review](https://github.com/openedx/openedx-k8s-harmony/pull/24#pullrequestreview-1409434403)
* [Updating commit messages](https://github.com/openedx/openedx-k8s-harmony/pull/24#issuecomment-1495904096) to match [Conventional Commits](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html#type) standards.
* Resolving merge conflicts.

Once that's done we can line this up for another review pass.

**Comment by lpm0073** [comment]
got it, thanks @itsjeyd. I hope to be able to take care of all three later today.

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/versions.tf]: i followed this recommendation.

File: infra-examples/aws/vpc/versions.tf
Comment (lpm0073) [reply infra-examples/aws/vpc/versions.tf:15]: i followed this recommendation.

File: infra-examples/aws/vpc/main.tf
Comment (lpm0073) [reply infra-examples/aws/vpc/main.tf]: i followed this recommendation

File: infra-examples/aws/main.tf
Comment (lpm0073) [reply infra-examples/aws/main.tf]: removed this module.

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:49]: i ignored this recommendation.

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:56]: i ignored this recommendation.

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:58]: i am pleased to report that said blank line has been removed :=)

File: infra-examples/aws/k8s-cluster/main.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/main.tf:53]: This is a good suggestion, but I don't have any experience implementing it. Maybe best if you create a new Issue for this, and we tackle this after this PR is merged?

File: infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/addon_ebs_csi_driver.tf:25]: i ignored this recommendation.

File: README.md
Comment (lpm0073) [reply README.md:242]: this is already merged via another PR. (But GREAT catch!!)

File: .gitignore
Comment (lpm0073) [reply .gitignore:13]: Ignored. Out of scope for this PR. This didn't come from me.

File: infra-examples/aws/k8s-cluster/main.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/main.tf]: i'm pleased to report that this blank line has been removed :)

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/versions.tf]: i followed this recommendation.

File: infra-examples/aws/k8s-cluster/versions.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/versions.tf]: i followed this recommendation.

File: infra-examples/aws/k8s-cluster/main.tf
Comment (lpm0073) [reply infra-examples/aws/k8s-cluster/main.tf]: done.

**Comment by lpm0073** [comment]
> @lpm0073 can you amend the commit messages to follow our conventional commit standard? https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html#type

editing these now.

**Comment by lpm0073** [comment]
> @lpm0073 It looks like there's three things to follow up on:
> 
> * Comments from @gabor-boros's [review](https://github.com/openedx/openedx-k8s-harmony/pull/24#pullrequestreview-1409434403)
> * [Updating commit messages](https://github.com/openedx/openedx-k8s-harmony/pull/24#issuecomment-1495904096) to match [Conventional Commits](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html#type) standards.
> * Resolving merge conflicts.
> 
> Once that's done we can line this up for another review pass.

these are done.

**Comment by lpm0073** [comment]
Thanks, that would be ideal.

**Comment by gabor-boros** [comment]
@itsjeyd Sure! I'm going to review this today to speed things up!

**Review by gabor-boros — APPROVED** [review]
@lpm0073 I just finished reviewing this. I think after a rebase onto latest master branch, this could be merged.

After rebasing, I'm going to test the changes by running terraform, so we can make sure the infra is created properly before merging, but since I'm not expecting any more change requests, I'm giving my approval here.

**Comment by lpm0073** [comment]
merge conflicts are resolved. Please merge this PR at your earliest convenience.

**Comment by lpm0073** [comment]
My commit message is improperly formatted 🤦

**Comment by gabor-boros** [comment]
I guess "feat: implement Karpenter example" would be a good squash commit message that should pass the test too

**Comment by gabor-boros** [comment]
Eh, kind of rewording solves the issue, but indeed unfair. I ran out of time today, but will take care of this this week 😇

**Comment by bradenmacdonald** [comment]
@gabor-boros friendly reminder to fix the commit msgs and merge this PR for @lpm0073  please :)

**Comment by jfavellar90** [comment]
@bradenmacdonald @lpm0073 could we just merge https://github.com/openedx/openedx-k8s-harmony/pull/40 before this one? Once https://github.com/openedx/openedx-k8s-harmony/pull/40 is merged, I will rebase and apply some changes to this PR in order to create a new Harmony chart release

**Comment by bradenmacdonald** [comment]
@jfavellar90 That's fine with me. Thanks for volunteering to do the rebase!

**Comment by openedx-webhooks** [comment]
@lpm0073 Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-k8s-harmony&entry.867055334=https://github.com/openedx/openedx-k8s-harmony/pull/24&entry.1484655318=https://github.com/lpm0073&entry.752974735=2023-03-06+13:08&entry.1917517419=2023-07-11+15:52&entry.2133058324=No) so we can improve your experience in the future.

**Pull request closed without merging**
